src/clawback_escrow/src/main.nr

```diff
@@ -0,0 +1,167 @@
+    }
+
+    #[private]
+    fn create_clawback_escrow(escrow: AztecAddress, receiver: AztecAddress, amount: U128) {
+        // This will revert if the Clawback Escrow is not the owner of the escrow 
+        let _ = Escrow::at(escrow).leak_keys(receiver);
+
+        let randomness = unsafe { random() };
+        let sender = context.msg_sender();
+        let mut note = ClawbackEscrowNote::new(
+            sender, 
+            receiver, 
+            escrow, 
+            amount,
+            randomness
+        );
+        
+        // Check that a clawback escrow has not been created for that escrow
+        let nullifier = std::hash::pedersen_hash([escrow.to_field()]); // derive nullifier from escrow address
+        context.push_nullifier(nullifier);
+
+        
+        let note_insertion = storage.escrows.insert(&mut note);
+
+        // Emit the note for the sender
+        note_insertion.emit(encode_and_encrypt_note(
+            &mut context,
+            sender,
+            sender,
+        ));
+
+        // Emit the note for the receiver
+        note_insertion.emit(encode_and_encrypt_note(
+            &mut context,
+            receiver,
+            sender,
+        ));
+    }
+
+    #[private]
+    fn claim(escrow: AztecAddress) {
+        let caller = context.msg_sender();
+        
+        // Find and remove the note
+        // There should only be one note per escrow
+        let note = _find_and_remove_clawback_note(storage, escrow);
+            
+        assert(note.escrow == escrow, "Invalid escrow address");
+        assert(caller == note.receiver, "Only receiver can claim");
+
+        Escrow::at(escrow).withdraw(AztecAddress::zero(), note.amount, note.receiver).call(&mut context);
+    }
+
+    #[private]
+    fn clawback(escrow: AztecAddress) {
+        let caller = context.msg_sender();
+        
+        // Find and remove the note
+        // There should only be one note per escrow
+        let note = _find_and_remove_clawback_note(storage, escrow);
+            
+        assert(note.escrow == escrow, "Invalid escrow address");
+        assert(caller == note.sender, "Only sender can clawback");
+
+        Escrow::at(escrow).withdraw(AztecAddress::zero(), note.amount, caller).call(&mut context);
+    }
+   
+    #[contract_library_method]
+    fn _find_and_remove_clawback_note(storage: Storage<&mut PrivateContext>, escrow: AztecAddress) -> ClawbackEscrowNote {
+        // let options = NoteGetterOptions::with_filter(filter_note_escrow, escrow);
+        let options = NoteGetterOptions::new()
+            .select(ClawbackEscrowNote::properties().escrow, Comparator.EQ, escrow)
+            .set_limit(1);
+        let notes = storage.escrows.pop_notes(options);
+
+        // We revert if there is no note in the filter so we can probably unwrap directly
+        let unwrapped_note = notes.get_unchecked(0);
+
+        assert(unwrapped_note.escrow == escrow, "Invalid escrow address");
+        unwrapped_note
+    }
```

Comment:
i'm debating if we should nullify the note or just fetch it (checking the preimage exists in the note-commitments tree), this particular use case seems to be 1-use, since is transferring the full amount to the one presenting the note.

in the case Alice clawbacks, Bob should see the note being nullified (not revert at the Token level because the BalanceNote was nullified), so i'd agree on fetching and nullifying it 👌

Reply:
addressing in #19